### DISCUSSION TITLE
Python freeopcua server: how to know number of connected clients?

### DISCUSSION THREAD

**Discussion by ryangu18** [FreeOpcUa/python-opcua#1469] [Q&A] [answered]
I have successfully set-up a freeopcua server and it communicates well with clients but now I would like to know how many clients are connected to my server. Is there a method or workaround to know how many clients are connected to a freeopcua server?

In fact, this is a question I found at https://stackoverflow.com/, I have the same requirement, I want to know how many connections the ua server currently has, and I want to know how to clean up invalid client connections, This will ensure that the currentSessionCount of ua will not be too large

Here is its answer, This answer disappointed me

``` For OPC UA it should be in Server/ServerDiagnostics/ServerDiagnosticsSummary/CurrentSessionCount

But afaik there is absolutely no code included about all this Diagnostics-Stuff neither in python-opcua nor in opcua-asyncio.
```

I would like to confirm that this answer. help me please. Thank you in advance.

**Comment by schroeder-** [comment, marked as answer]
This should also work for python-opcua: https://github.com/FreeOpcUa/opcua-asyncio/discussions/845

**Reply by ryangu18** [reply]
Thank you very much for your help, through
` connectInfo = len(server.bserver.clients) `

I get the number of connections.

This attributes in the UA server represents the number of connections
![image](https://user-images.githubusercontent.com/99937589/171321064-06e2a5bc-7063-49c0-8cea-46a29ad249f0.png)

However, in the ua server created by freeopcua library, this is blank. 
![image](https://user-images.githubusercontent.com/99937589/171321245-b64f0765-8fa9-4825-b308-f4b220658307.png)

How to write the obtained connection count( len(server.bserver.clients) )  to there (i=2277) ?

**Reply by ryangu18** [reply]
This is the supplemental information， I just used
`server.get_node("i=2277").set_writable()`
and 
`client.get_node("i=2277").set_value(connectInfo)`
The writing of data is implemented, but is this a norm (best practice) practice?

If yes, please allow me to shift focus to the second question,
``` and I want to know how to clean up invalid client connections, This will ensure that the currentSessionCount of ua will not be too large ```

**Reply by schroeder-** [reply]
Normaly you would reduce the SessionTimeout for the clients if you have a problem with dead sessions. But there is currently no possiblity to reduce this on the server side.

**Reply by ryangu18** [reply]
Sincerely thank you for your help, maybe this problem is a bit complicated.

```
from opcua import Client
client = Client("opc.tcp://192.168.32.239:49320")
client.connect()
while True:  # read data loop
   try:
       """
       read (or write) data from (to) UA server
       """
   except Exception as e1:  #The network interruption will cause the read/write to fail and enter the except instruction
       	try:
		client.disconnect()  # Its execution is invalid when the network is down
	except:
		pass
	client.connect()
        sleep(10)
```

In this sample code, after a client connects to the server I created, it will keep reading data, but when the network is interrupted, it will reconnect, please note that the client's previous connection cannot be closed (because the network has been interrupted). So now the number of server connections is 2, an old connection and a new connection that is re-established now. So I think the server is needed to clean up invalid connections, not the client's function.
I have done a similar experiment in kepserver. When the connection is interrupted, the client initiates a reconnection. In the kepserver server, I see that the number of connections = 2. However, after a few minutes, the currentSessionCount of the kepserver server returns to 1. So I think this is a mechanism for the kepserver server to periodically clean up invalid connections.
This is 5.6.2 of the OPC UA Online Reference. https://reference.opcfoundation.org/v104/Core/docs/Part4/5.6.2/
```A Server application should limit the number of Sessions. To protect against misbehaving Clients and denial of service attacks, the Server shall close the oldest Session that is not activated before reaching the maximum number of supported Sessions. ```

I hope that the ua server created with the freeopcua library has similar functions, so as to avoid network interruptions that cause the currentSessionCount to continue to increase and affect the server efficiency. But I don't know how should I implement this idea, some command or some parameter？


But I'm new to this, if there is a better way, or if I'm misunderstanding it, please don't hesitate to point it out to me.
I'm very sorry to ask you such a complicated question, but it's been bothering me for a long time. Whether I got your help or not, I would like to thank you in advance

**Reply by schroeder-** [reply]
First of the serverside of python-opcua is a bit barebone with limiting resources, there is no limit for sessions and so on.

Second the implementation is that the server and client negotiate a session timeout, a session in theory i outlives a tcp connection, but the server has no ability to restore sessions.

If you controll the client, you can set a lower timeout.
`client = Client("opc.tcp://192.168.0.1:4840", timeout=360000)`
Timeout is in milliseconds, try something like 5000 for 5s.

**Reply by ryangu18** [reply]
Due to the limitations of this library, some functions cannot be implemented, which I regret. But I've got the answer I need. Thanks for your help, I closed this discussion.

**Comment by shineworld** [comment]
Very bad programming way, which breaks any OOP rule, but you can do this:

`current_connections = getattr(self.server.iserver.isession, "_current_connections", 0)`
where self.server is Server() class from:
`from asyncua import Server`

The class InteralSession keep the protected _current_connections counter that, at moment 1.0pre, is
accessible only using getattr() with a default result of 0 if the attribute will be removed in future versions.


In my case, I have a server that exposes a node for server debugging and the number of connected sessions is important:

```
            # creates folder and nodes for server info
            if self.show_server_info_node:
                f_server_info = await o_rosetta_cnc.add_folder(idx, "ServerInfo")
                v_time_1 = await node_add_variable(f_server_info, "time_1", "", 0.0, datatype=vt_Double, varianttype=id_Double, writable=False)
                v_time_2 = await node_add_variable(f_server_info, "time_2", "", 0.0, datatype=vt_Double, varianttype=id_Double, writable=False)
                v_sessions = await node_add_variable(f_server_info, "sessions", "", 0, datatype=vt_Int32, varianttype=id_Int32, writable=False)

        # server update management
        async with self.server:

            # server updating loop
            while True:
                   ...
                   ...
                   # updates servers info nodes values
                    if self.show_server_info_node:
                        await v_time_1.write_value(t3 - t1, vt_Double)
                        await v_time_2.write_value(t3 - t2, vt_Double)
                        current_connections = getattr(self.server.iserver.isession, "_current_connections", 0)
                        await v_sessions.write_value(current_connections, vt_Int32)

```
